File: backend/app/controllers/society_controller.py
from flask import request
from app.database.mongodb import mongo
from app.utils.responses import success_response, error_response
import datetime
import uuid

class SocietyController:
    @staticmethod
    def register_society():
        print("DEBUG: Society Registration request received")
        data = request.get_json()
        print(f"DEBUG: Payload: {data}")
        name = data.get('society_name')
        city = data.get('city')
        address = data.get('address')
        total_flats = data.get('total_flats')
        admin_email = data.get('admin_email')
        
        if not all([name, city, admin_email]):
            return error_response("Missing required fields")
            
        # Check if society name already exists
        if mongo.db['societies'].find_one({"society_name": name}):
            return error_response("Society with this name already exists")
            
        # Generate unique society code
        society_code = str(uuid.uuid4())[:8].upper()
        
        society_id = mongo.db['societies'].insert_one({
            "society_name": name,
            "society_code": society_code,
            "city": city,
            "address": address,
            "total_flats": total_flats,
            "is_active": True,
            "subscription_plan": "Free Trial",
            "created_at": datetime.datetime.utcnow()
        }).inserted_id
        
        # Initialize Society Wallet
        mongo.db['society_wallets'].insert_one({
            "society_id": str(society_id),
            "balance": 0.0,
            "last_updated": datetime.datetime.utcnow()
        })
        
        return success_response("Society registered successfully", {
            "society_id": str(society_id),
            "society_code": society_code
        }, 201)

    @staticmethod
    def get_society_by_code(code):
        society = mongo.db['societies'].find_one({"society_code": code.upper()})
        if not society:
            return error_response("Invalid society code", 404)
            
        return success_response("Society found", {
            "society_id": str(society['_id']),
            "society_name": society['society_name']
        })

    @staticmethod
    def get_all_societies():
        societies = list(mongo.db['societies'].find())
        for s in societies:
            s['_id'] = str(s['_id'])
        return success_response("Societies fetched", societies)

File: backend/src/controllers/society.controller.js
import Society from '../models/Society.js';
import User from '../models/User.js';
import Flat from '../models/Flat.js';
import ApiError from '../utils/ApiError.js';
import ApiResponse from '../utils/ApiResponse.js';
import asyncHandler from '../utils/asyncHandler.js';
import { generateSocietyCode } from '../utils/helpers.js';
import slugify from 'slugify';

class SocietyController {
  /**
   * POST /api/v1/societies/register
   * Public — registers a new society and creates admin account.
   */
  static register = asyncHandler(async (req, res) => {
    const { societyName, city, state, pincode, address, totalFlats, totalBlocks, adminName, adminEmail, adminPassword, adminPhone } = req.body;

    // Check if admin email already exists
    const existingUser = await User.findOne({ email: adminEmail });
    if (existingUser) {
      throw ApiError.conflict('An account with this admin email already exists.');
    }

    // Generate unique code
    let code;
    let codeExists = true;
    while (codeExists) {
      code = generateSocietyCode();
      codeExists = await Society.findOne({ code });
    }

    // Create admin user first
    const admin = await User.create({
      fullName: adminName,
      email: adminEmail,
      password: adminPassword,
      phone: adminPhone,
      role: 'SOCIETY_ADMIN',
      authProvider: 'local',
      isVerified: true,
    });

    // Create society
    const society = await Society.create({
      name: societyName,
      code,
      slug: slugify(societyName, { lower: true, strict: true }),
      address: {
        line1: address,
        city,
        state: state || city,
        pincode: pincode || '000000',
      },
      totalFlats: parseInt(totalFlats, 10) || 50,
      totalBlocks: parseInt(totalBlocks, 10) || 1,
      adminId: admin._id,
      contactEmail: adminEmail,
      contactPhone: adminPhone,
    });

    // Link society to admin
    admin.societyId = society._id;
    await admin.save();

    // Auto-generate flat entries
    const flatDocs = [];
    for (let i = 1; i <= Math.min(society.totalFlats, 500); i++) {
      flatDocs.push({
        societyId: society._id,
        flatNumber: `${i}`,
        block: 'A',
        floor: Math.ceil(i / 4),
      });
    }
    if (flatDocs.length > 0) {
      await Flat.insertMany(flatDocs, { ordered: false }).catch(() => {});
    }

    ApiResponse.created('Society registered successfully', {
      society: {
        id: society._id,
        name: society.name,
        code: society.code,
        totalFlats: society.totalFlats,
      },
      admin: {
        id: admin._id,
        email: admin.email,
      },
    }).send(res);
  });

  /**
   * GET /api/v1/societies/code/:code
   * Public — verify a society exists by code.
   */
  static getByCode = asyncHandler(async (req, res) => {
    const society = await Society.findOne({ code: req.params.code.toUpperCase() })
      .select('name code address totalFlats status');

    if (!society) {
      throw ApiError.notFound('No society found with this code.');
    }

    ApiResponse.ok('Society found', society).send(res);
  });

  /**
   * GET /api/v1/societies/:id
   * Protected — get society details.
   */
  static getById = asyncHandler(async (req, res) => {
    const society = await Society.findById(req.params.id)
      .populate('adminId', 'fullName email')
      .populate('committeeMembers.userId', 'fullName email');

    if (!society) {
      throw ApiError.notFound('Society not found.');
    }

    ApiResponse.ok('Society details', society).send(res);
  });

  /**
   * GET /api/v1/societies
   * Super Admin — list all societies.
   */
  static getAll = asyncHandler(async (req, res) => {
    const { page = 1, limit = 20, status, search } = req.query;

    const filter = {};
    if (status) filter.status = status;
    if (search) {
      filter.$or = [
        { name: { $regex: search, $options: 'i' } },
        { code: { $regex: search, $options: 'i' } },
      ];
    }

    const societies = await Society.find(filter)
      .sort({ createdAt: -1 })
      .skip((page - 1) * limit)
      .limit(parseInt(limit, 10))
      .populate('adminId', 'fullName email');

    const total = await Society.countDocuments(filter);

    ApiResponse.ok('Societies list', {
      societies,
      pagination: {
        page: parseInt(page, 10),
        limit: parseInt(limit, 10),
        total,
        pages: Math.ceil(total / limit),
      },
    }).send(res);
  });

  /**
   * PATCH /api/v1/societies/:id/status
   * Super Admin — activate/suspend a society.
   */
  static updateStatus = asyncHandler(async (req, res) => {
    const { status } = req.body;
    
    const society = await Society.findByIdAndUpdate(
      req.params.id,
      { status },
      { new: true, runValidators: true }
    );

    if (!society) {
      throw ApiError.notFound('Society not found.');
    }

    ApiResponse.ok(`Society ${status.toLowerCase()} successfully`, society).send(res);
  });
}

export default SocietyController;
